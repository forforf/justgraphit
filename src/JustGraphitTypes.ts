
// Not perfect, but ok for now. Maybe improve if typescript supports regexes
export type ISODatetimeType = `${number}-${number}-${number}T${number}:${number}:${number}.${number}Z`;

export type GraphName = string;

export type DateOrIso = Date | ISODatetimeType;

// individual data points
export type JustGraphitEntry = {
  number: number;
  datetime: ISODatetimeType;
  dt?: Date;
}

export type JustGraphitStore = Record<GraphName, JustGraphitEntry[]>

export type PassthroughFunction = (...args: unknown[]) => unknown;

export type RefSvg = SVGSVGElement | null;

export type D3Select = (ref: RefSvg) => PassthroughFunction;

// Functions that handle graphs by graphName (changing graphs, deleting graphs, etc)
export type GraphHandler = (graphName: GraphName) => void;

export type GraphEntryHandler = (graphName: GraphName, numberInput: string) => void;