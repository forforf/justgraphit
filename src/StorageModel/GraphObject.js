
class GraphObject {

  static fromPlainObject(obj) {
    const ks = Object.keys(obj);
    if (ks.length > 1) {
      throw new Error(`Only single keyed object supported, but got ${ks}`)
    }
    const name = Object.keys(obj)[0];
    const data = obj[name];
    return new GraphObject(name, data);
  }

  constructor(graphName, graphData) {
    this.name = graphName;
    this.data = graphData;
  }
}

export default GraphObject