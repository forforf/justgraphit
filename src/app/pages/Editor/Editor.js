
import React, { Component } from 'react';
import Edit from 'EditorView/Edit';

class Editor extends Component {
  
  render() {
    return (
        <Edit storage={this.props.storage}/>
    );
  }
}


export default Editor;