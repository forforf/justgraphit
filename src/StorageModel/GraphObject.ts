import {GraphName, JustGraphitEntry, JustGraphitStore} from "../JustGraphitTypes";


// Used to enforce that there is a single key/value pair (key = name, value = data in this case)
class GraphObject {
  name: GraphName;
  data: JustGraphitEntry[];

  // input object should look like a store, but actually only have a single graph -> data entry
  static fromPlainObject(obj: JustGraphitStore): GraphObject {
    const ks = Object.keys(obj);
    if (ks.length > 1) {
      throw new Error(`Only single keyed object supported, but got ${ks}`)
    }
    const name = Object.keys(obj)[0];
    const data = obj[name];
    return new GraphObject(name, data);
  }

  constructor(graphName: GraphName, graphData: JustGraphitEntry[]) {
    this.name = graphName;
    this.data = graphData;
  }
}

export default GraphObject;