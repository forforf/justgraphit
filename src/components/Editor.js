import React, {useState} from 'react';
import DeleteIcon from '@mui/icons-material/Delete';
import GraphObject from '../StorageModel/GraphObject';
import {DataGrid, GridToolbar} from '@mui/x-data-grid';
import './Editor/Editor.css';

// Wrap the Delete Icon in the Datagrid classes for styling.
// Think about a better approach if this function is revisited in the future.
function DgDeleteIcon({className:passedClasses, ...props}) {
  return (
    <div className="MuiDataGrid-toolbarContainer">
      <div className="MuiButton-textPrimary MuiButton-textSizeSmall">
        <DeleteIcon {...props} />
      </div>
    </div>
  )
}

function CustomToolbar({deleteHandler}) {
  return (
    <div style={{display:"flex", justifyContent:"space-between"}}>
      <GridToolbar />
      <DgDeleteIcon  onClick={deleteHandler}/>
    </div>
  );
}

// Removes rows based on the provided array of datetimes.
function removeByDatetime(removeDatetimes, rows) {
  const removeSet = new Set(removeDatetimes);
  return rows.filter(r => !removeSet.has(r.datetime));
}

// TODO: Should `setGraphObject`  save to storage?, or should it be done here? Leaning toward here
//   Reason: Confirmations and potential transformations, might make rolling back easier in the future.
//   Counter: Why should this component know what a graphObject is?
function Editor({ graphObject, setGraphObject, saveToStorage }) {

  // state
  const [selectionModel, setSelectionModel] = useState([]);

  const align = { headerAlign: 'center', align: 'center'};
  const editable = { editable: true}
  const columns = [
    { field: 'datetime', headerName: 'Datetime Stamp', width: 220,...editable, ...align },
    { field: 'number', headerName: 'Value', width: 150, ...editable, ...align}
  ];
  const rows = graphObject.data;


  const deleteHandler = () => {
    if(window.confirm(`Delete ${selectionModel.length} rows?`)) {
      const remainingRows = removeByDatetime(selectionModel, rows);
      const newGraphObject = new GraphObject(graphObject.name, remainingRows);
      setGraphObject(newGraphObject);
      saveToStorage(graphObject.name, remainingRows);
    }
  }

  // Since the toolbar is passed as a bare component
  // Use a HOC leveraging closure to add our delete handler
  const withDeleteHandler = Component => ({...props}) => (<Component deleteHandler={deleteHandler} {...props}/>);
  const ToolBarWrapper = withDeleteHandler(CustomToolbar);

  return (
    // TODO: Figure out where to style (theme, css, but probably not here)
    <div style={{ height: 600, width: '90%', margin: 'auto' }}>
      <h3>{graphObject.name}</h3>
      <DataGrid
        rows={rows}
        getRowId={r => r.datetime}
        columns={columns}
        pageSize={20}
        rowsPerPageOptions={[20]}
        checkboxSelection
        disableSelectionOnClick
        onSelectionModelChange={setSelectionModel}
        components={{
          Toolbar: ToolBarWrapper,
        }}
      />
    </div>
  )
}

export default Editor;
