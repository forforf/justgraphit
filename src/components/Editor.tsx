import React, { useState } from 'react';
import DeleteIcon from '@mui/icons-material/Delete';
import GraphObject from '../StorageModel/GraphObject';
import {
  GraphName,
  ISODatetimeType,
  JustGraphitEntry,
} from '../JustGraphitTypes';
import {
  DataGrid,
  GridAlignment,
  GridColDef,
  GridRowModel,
  GridSelectionModel,
  GridToolbar,
} from '@mui/x-data-grid';
import './Editor/Editor.css';

// Wrap the Delete Icon in the Datagrid classes for styling.
// Think about a better approach if this function is revisited in the future.
function DgDeleteIcon({ ...props }) {
  return (
    <div className="MuiDataGrid-toolbarContainer">
      <div className="MuiButton-textPrimary MuiButton-textSizeSmall">
        <DeleteIcon {...props} />
      </div>
    </div>
  );
}

type CustomToolbarProps = {
  deleteHandler: () => void;
};

function CustomToolbar({ deleteHandler }: CustomToolbarProps) {
  return (
    <div style={{ display: 'flex', justifyContent: 'space-between' }}>
      <GridToolbar />
      <DgDeleteIcon onClick={deleteHandler} />
    </div>
  );
}

// Removes rows based on the provided array of datetimes.
function removeByDatetime(
  removeDatetimes: ISODatetimeType[],
  rows: JustGraphitEntry[]
) {
  const removeSet = new Set(removeDatetimes);
  return rows.filter((r) => !removeSet.has(r.datetime));
}

export type EditorProps = {
  graphObject: GraphObject;
  setGraphObject: (g: GraphObject) => void;
  saveToStorage: (gn: GraphName, r: JustGraphitEntry[]) => void;
};
// TODO: Should `setGraphObject` save to storage internally?,
//  or should saving to storage be done here? Leaning toward here
//   Reason: Confirmations and potential transformations, might make rolling back easier in the future.
//   Counter: Why should this component know what a graphObject is?
function Editor({
  graphObject,
  setGraphObject,
  saveToStorage,
}: EditorProps): JSX.Element {
  // state
  const [selectionModel, setSelectionModel] = useState<GridSelectionModel>([]);

  const align = {
    headerAlign: 'center' as GridAlignment,
    align: 'center' as GridAlignment,
  };
  const editable = { editable: true };
  const columns: GridColDef[] = [
    {
      field: 'datetime',
      headerName: 'Datetime Stamp',
      width: 220,
      ...editable,
      ...align,
    },
    { field: 'number', headerName: 'Value', width: 150, ...editable, ...align },
  ];
  const rows = graphObject.data;

  const deleteHandler = () => {
    if (window.confirm(`Delete ${selectionModel.length} rows?`)) {
      const remainingRows = removeByDatetime(
        selectionModel as ISODatetimeType[],
        rows
      );
      const newGraphObject = new GraphObject(graphObject.name, remainingRows);
      setGraphObject(newGraphObject);
      saveToStorage(graphObject.name, remainingRows);
    }
  };

  const processRowUpdate = React.useCallback(
      (newRow: GridRowModel) => {
        console.log('nr', newRow)
        // TODO: Dry with deleteHandler and see if it can be simplified
        const newGraphRows = rows.map( r => {
          if (r.datetime == newRow.datetime) {
            const {number, datetime, dt} = newRow
            return {number, datetime, dt}
          }
          return r
        })
        const newGraphObject = new GraphObject(graphObject.name, newGraphRows);
        setGraphObject(newGraphObject);
        saveToStorage(graphObject.name, newGraphRows);
        return newRow
      },
      [graphObject.name, rows, saveToStorage, setGraphObject],
  );

  const handleProcessRowUpdateError = React.useCallback((error: Error) => {
    console.error(error);
  }, []);

  type WithDeleteProps = {
    deleteHandler: () => void;
  };

  // Since the toolbar is passed as a bare component (no props)
  // Use a HOC leveraging closure to add our delete handler
  const withDeleteHandler = function ComponentWrapper(
    Component: React.FC<WithDeleteProps>
  ) {
    return function WithDeleteHandler({ ...props }) {
      return <Component deleteHandler={deleteHandler} {...props} />;
    };
  };
  const ToolBarWrapper = withDeleteHandler(CustomToolbar);

  return (
    // TODO: Figure out where to style (theme, css, but probably not here)
    <div style={{ height: 600, width: '90%', margin: 'auto' }}>
      <h3>{graphObject.name}</h3>
      <DataGrid experimentalFeatures={{ newEditingApi: true }}
        rows={rows}
        getRowId={(r) => r.datetime}
        columns={columns}
        pageSize={20}
        rowsPerPageOptions={[20]}
        checkboxSelection
        disableSelectionOnClick
        onSelectionModelChange={setSelectionModel}
        processRowUpdate={processRowUpdate}
        onProcessRowUpdateError={handleProcessRowUpdateError}
        components={{
          Toolbar: ToolBarWrapper,
        }}
      />
    </div>
  );
}

export default Editor;
