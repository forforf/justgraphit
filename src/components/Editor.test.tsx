import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import Editor, { EditorProps } from './Editor';
import GraphObject from '../StorageModel/GraphObject';
import { ISODatetimeType, JustGraphitEntry } from '../JustGraphitTypes';

let editorProps: EditorProps;
beforeEach(() => {
  const graphData: JustGraphitEntry[] = [
    {
      number: 1,
      datetime: new Date(
        '1999-12-31T23:59:59.00Z'
      ).toISOString() as ISODatetimeType,
    },
    {
      number: 2,
      datetime: new Date(
        '2000-01-02T00:00:00.000Z'
      ).toISOString() as ISODatetimeType,
    },
  ];
  editorProps = {
    graphObject: new GraphObject('a graph', graphData),
    setGraphObject: jest.fn(),
    saveToStorage: jest.fn(),
  };
});

test('deep snapshot', () => {
  const component = renderer.create(<Editor {...editorProps} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('shallow snapshot', () => {
  const renderer = ShallowRenderer.createRenderer();
  renderer.render(<Editor {...editorProps} />);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
