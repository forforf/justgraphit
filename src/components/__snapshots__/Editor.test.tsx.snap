// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep snapshot 1`] = `
<div
  style={
    Object {
      "height": 600,
      "margin": "auto",
      "width": "90%",
    }
  }
>
  <h3>
    a graph
  </h3>
</div>
`;

exports[`shallow snapshot 1`] = `
<div
  style={
    {
      "height": 600,
      "margin": "auto",
      "width": "90%",
    }
  }
>
  <h3>
    a graph
  </h3>
  <Memo
    checkboxSelection={true}
    columns={
      [
        {
          "align": "center",
          "editable": true,
          "field": "datetime",
          "headerAlign": "center",
          "headerName": "Datetime Stamp",
          "width": 220,
        },
        {
          "align": "center",
          "editable": true,
          "field": "number",
          "headerAlign": "center",
          "headerName": "Value",
          "width": 150,
        },
      ]
    }
    components={
      {
        "Toolbar": [Function],
      }
    }
    disableSelectionOnClick={true}
    getRowId={[Function]}
    onSelectionModelChange={[Function]}
    pageSize={20}
    rows={
      [
        {
          "datetime": "1999-12-31T23:59:59.000Z",
          "number": 1,
        },
        {
          "datetime": "2000-01-02T00:00:00.000Z",
          "number": 2,
        },
      ]
    }
    rowsPerPageOptions={
      [
        20,
      ]
    }
  />
</div>
`;
